### PR TITLE
implement uniswap resolver

### PR DESCRIPTION
**Pull request by cryptoDev222**
Implement Uniswap V3 resolver.
Get liquidity amount from deposit.
Get amountA, amountB from withdraw & collect.

### CODE REVIEW

File: contracts/protocols/mainnet/uniswap/helpers.sol [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+        if (sqrtRatioX96 <= sqrtRatioAX96) {
+            amount0 = getAmount0ForLiquidity(sqrtRatioAX96, sqrtRatioBX96, liquidity);
+        } else if (sqrtRatioX96 < sqrtRatioBX96) {
+            amount0 = getAmount0ForLiquidity(sqrtRatioX96, sqrtRatioBX96, liquidity);
+            amount1 = getAmount1ForLiquidity(sqrtRatioAX96, sqrtRatioX96, liquidity);
+        } else {
+            amount1 = getAmount1ForLiquidity(sqrtRatioAX96, sqrtRatioBX96, liquidity);
+        }
+    }
+}
```
Comment (thrilok209) [inline contracts/protocols/mainnet/uniswap/helpers.sol]: Maybe we can create another lib contract in libraries file? https://github.com/Uniswap/uniswap-v3-periphery/blob/main/contracts/libraries/LiquidityAmounts.sol

File: contracts/protocols/mainnet/uniswap/helpers.sol [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+                false
+            );
+        } else {
+            amount1 = SqrtPriceMath.getAmount1Delta(
+                TickMath.getSqrtRatioAtTick(tickLower),
+                TickMath.getSqrtRatioAtTick(tickUpper),
+                liquidity,
+                false
+            );
+        }
```
Comment (thrilok209) [inline contracts/protocols/mainnet/uniswap/helpers.sol]: @cryptoDev222 I think we can use this function: https://github.com/Uniswap/uniswap-v3-periphery/blob/1f5b385675eee30a77ee10f4dfd6aa0aaa766026/contracts/libraries/LiquidityAmounts.sol#L120

File: contracts/protocols/mainnet/uniswap/helpers.sol [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+            uint160 sqrtRatioBX96 = TickMath.getSqrtRatioAtTick(tickUpper);
+
+            liquidity = LiquidityAmounts.getLiquidityForAmounts(
+                sqrtPriceX96,
+                sqrtRatioAX96,
+                sqrtRatioBX96,
+                amountA,
+                amountB
+            );
+        }
```
Comment (thrilok209) [inline contracts/protocols/mainnet/uniswap/helpers.sol:293]: @cryptoDev222 After this, we need to check at what ratio amountA and amountB is required. So, we need to pass `liquidity` into this function: https://github.com/Uniswap/uniswap-v3-periphery/blob/1f5b385675eee30a77ee10f4dfd6aa0aaa766026/contracts/libraries/LiquidityAmounts.sol#L120

To get `amountA` and `amountB`

File: contracts/protocols/mainnet/uniswap/helpers.sol (hunk 1 above)
Comment (cryptoDev222) [reply contracts/protocols/mainnet/uniswap/helpers.sol:110]: Okay, got it.

File: contracts/protocols/mainnet/uniswap/helpers.sol (hunk 3 above)
Comment (cryptoDev222) [reply contracts/protocols/mainnet/uniswap/helpers.sol:261]: You mean something like this.
https://github.com/Instadapp/dsa-resolvers-old/blob/master/contracts/protocols/mainnet/uniswapV2.sol#L356

File: contracts/protocols/mainnet/uniswap/helpers.sol [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+     */
+    function getLastNftId(address user) internal view returns (uint256 tokenId) {
+        uint256 len = nftManager.balanceOf(user);
+        tokenId = nftManager.tokenOfOwnerByIndex(user, len - 1);
+    }
+
+    function userNfts(address user) internal view returns (uint256[] memory tokenIds) {
+        uint256 len = nftManager.balanceOf(user);
+        for (uint256 i = 0; i < len; i++) {
+            uint256 tokenId = nftManager.tokenOfOwnerByIndex(user, len - 1);
```
Comment (thrilok209) [inline contracts/protocols/mainnet/uniswap/helpers.sol]:
```suggestion
            uint256 tokenId = nftManager.tokenOfOwnerByIndex(user, i);
```

File: contracts/protocols/mainnet/uniswap/helpers.sol [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+            amount0 = sub(amountA, amount0);
+            amount1 = sub(amountB, amount1);
+        }
+    }
+
+    function singleDepositAmount(
+        uint256 tokenId,
+        address tokenA,
+        uint256 amountA
+    ) internal view returns (address tokenB, uint256 amountB) {
```
Comment (thrilok209) [inline contracts/protocols/mainnet/uniswap/helpers.sol]: @cryptoDev222 We also need to return the correct order of the token addresses and amounts.

File: contracts/protocols/mainnet/uniswap/main.sol [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+            uint24 fee,
+            int24 tickLower,
+            int24 tickUpper,
+            uint128 liquidity
+        )
+    {
+        (token0, token1, fee, tickLower, tickUpper, liquidity) = positions(tokenId);
+    }
+
+    struct PositionInfo {
```
Comment (thrilok209) [inline contracts/protocols/mainnet/uniswap/main.sol]: @cryptoDev222 can we also return `poolAddress`, `tokensOwed0`, `tokensOwed1`, `token1Amount`, `token0Amount`, `currentTick`?

File: contracts/protocols/mainnet/uniswap/main.sol [hunk 7]
```diff
@@ -97,6 +97,33 @@ contract Resolver is Helpers {
         (liquidity, tokenB, amountB, amountAMin, amountBMin) = singleDepositAmount(tokenId, tokenA, amountA, slippage);
     }
 
+    function getSingleMintAmount(
+        address tokenA,
+        uint256 amountA,
+        address tokenB,
+        uint256 slippage,
+        int24 tickLower,
+        int24 tickUpper
```
Comment (thrilok209) [inline contracts/protocols/mainnet/uniswap/main.sol]: @cryptoDev222 I think we need to pass `fee`?

**Review by KaymasJain — APPROVED** [review]

**Pull request merged**